### DISCUSSION TITLE
How to customize default controls?

### DISCUSSION THREAD

**Discussion by SmauDistribution** [AvaloniaUI/Avalonia#14515] [Q&A] [answered]
Hello everyone!
I would need to modify the background of the expander's header, but I don't know how to do it... I had a similar issue with the textbox... I needed to change the background during hover and I solved it with this code:

`<Style Selector="TextBox:pointerover /template/ Border">
 <Setter Property="Background" Value="Transparent"></Setter>
</Style>`

However I'm not quite clear on why I have to write /template/ Border... I've also tried searching various forums for a solution... but I'm still unclear on how to proceed with customizing the controls, since I've opened this discussion, I think it might be helpful for everyone to have a brief general explanation of how to customize the default controls.

Thanks in advance.

**Comment by timunie** [comment, marked as answer]
that is a really frequent question here. The solution I use is to check out original styles in source to get the selector to use. 

https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Controls/Expander.xaml

**Reply by Shadowblitz16** [reply]
Why don't you guys improve the framework then so you don't get them?
For one you guys expect people to magically know how the glue works and what controls have what parts.

I would suggest added a property editor and a scene tree to your vs, vscode and rider plugins so people don't have to read your mind or look up existing themes and know how they work.

**Reply by thevortexcloud** [reply]
> Why don't you guys improve the framework then so you don't get them?
> For one you guys expect people to magically know how the glue works and what 

They already did that (I am guessing because [I also complained about this](https://github.com/AvaloniaUI/Avalonia/discussions/14131#discussioncomment-8045181) in the past, and it was fixed ~1 week later). You now get compiler warnings/errors if your control template is missing a required part. 

https://github.com/AvaloniaUI/Avalonia/pull/14180

The only thing they have not done is automatically generated the list of required parts for docs. Which could be done fairly easily since all the information is already there. But it needs some infrastructure work on the docs side. So the easiest way to see the list of required parts is to still look at the source code.

> I would suggest added a property editor and a scene tree

This already exists? It's the dev tools which you open by pressing F12.

**Comment by SmauDistribution** [comment]
Great! I've solved it using this code:

`<Style Selector="Expander /template/ ToggleButton">
	<Setter Property="Background" Value="Transparent"></Setter>
</Style>`

Before closing the discussion, it might make sense to explain to others, especially newcomers like me, the structure behind the controls, or perhaps provide sources that discuss this. Anyway, thanks for the response!

**Reply by thevortexcloud** [reply]
https://docs.avaloniaui.net/docs/concepts/templated-controls

The way the control templates work is pretty much the same as how they work in WPF.  WPF's documentation is somewhat more detailed than Avalonia's about this. So you can give that a read too:

https://learn.microsoft.com/en-us/dotnet/desktop/wpf/controls/how-to-create-apply-template?view=netdesktop-8.0
https://learn.microsoft.com/en-us/dotnet/api/system.windows.controls.controltemplate?view=windowsdesktop-8.0

They are "lookless" which means they can be customised to look like anything. They contain template parts (which are controls that start with "PART_" in their name) which are typically mandatory components of the control. Apart from that there is no way to guarantee what a control template contains.

**Reply by SmauDistribution** [reply]
Great, thank you!
